python/tskit/tables.py

```diff
@@ -3020,6 +3026,99 @@ def union(
                 record=json.dumps(provenance.get_provenance_dict(parameters))
             )
 
+    def time_join(self, recent, time, continue_nodes=None):
+        ancient_ts = self.tree_sequence()
+        recent_ts = recent.tree_sequence()
+        if continue_nodes is None:
+            continue_nodes = ancient_ts.samples()
+
+        new_tables = recent_ts.tables
+        new_nodes = np.empty(0, dtype=np.uint32)
+        for t in recent_ts.trees():
+            for root_id in t.roots:
+                root = recent_ts.node(root_id)
+                if root.time == time:
+                    new_nodes = np.append(new_nodes, root_id)
+                elif root.time < time:
+                    n = new_tables.nodes.add_row(time=time)
+                    new_tables.edges.add_row(
+                        left=t.interval.left,
+                        right=t.interval.right,
+                        parent=n,
+                        child=root_id,
+                    )
+                    new_nodes = np.append(new_nodes, n)
+                else:
+                    raise RuntimeError(
+                        f"Root found older than join time (age = {root.time})"
+                    )
+        recent_ts = new_tables.tree_sequence()
+
+        if len(new_nodes) > len(continue_nodes):
+            raise RuntimeError(
+                f"""More lineages present at the longest-ago end of `recent_ts` \
+                than there are nodes in `continue_nodes`. \
+                Setting sample_size={len(continue_nodes)} or leaving it to default \
+                will ensure compliance."""
+            )
+
+        node_map = np.repeat(tskit.NULL, recent_ts.num_nodes)
+
+        node_map[new_nodes] = np.random.choice(
```

Comment:
This needs a random seed to be set to make it reproducible. We should add a ``random_seed=None`` argument  to the ``time_join`` function, and allocate an instance of ``np.random.Random``.